docs/runtime-user-guide.md

```diff
@@ -0,0 +1,119 @@
+# Runtime User Guide
+
+This document is focused on clarifying the expectations of the user
+container running in a Knative environment. This is separate from the expectations of the
+environment itself. The expectations of the environment are detailed in the [runtime
+contract](https://github.com/knative/serving/blob/master/docs/runtime-contract.md).
+
+The target audience of this document are _developers_ and _language and tooling
+developers_ as defined below:
+
+- **Developers** write code which is packaged into a container which is run on
+  the Knative cluster.
+- **Language and tooling developers** typically write tools used by
+    _developers_ to package code into containers. As such, they are concerned
+    that tooling which wraps developer code complies with this runtime contract.
+
+## General Behavior
+
+Containers running in Knative should be stateless containers. Stateless
+containers should have the following properties:
+
+- Fast startup time: The time from container launch to serving traffic should be
+  minmized to enable the platform to provide rapid autoscaling and scale-to-zero
+  semantics.
+- Minimize local state: The container may be killed and replaced to support
+  autoscaling and cluster maintenance operations. The performance or correctness
+  of the application should not rely on state persisted locally.
+- CPU usage only while requests are active: As containers may be killed or
+  suspended when no connections are active the container should only perform
+  operations that complete synchronously with the request's response.
+
+## Container Termination
+
+Knative containers should respond to `SIGTERM` signals to clean-up any resources
+necessary for a graceful shutdown.
+
+## Connection
+
+A Knative container must package a webserver that can respond to HTTP/1.1
+requests on a specified port.
```

Comment:
What about http/2?

Reply:
HTTP/2.0 is covered in the `Protocols` section. 

My understanding is that we expect clients to implement `h2c` and not `h2` which requires an initial upgrade request over HTTP/1.1. This is why I have set the expectation that responding to an HTTP/1.1 request is required.

Reply:
It might be worth clarifying that containers can implement h2c with the logic you provided here.

Reply:
Makes sense. I can add this.